tempesta_fw/apm.c

```diff
@@ -0,0 +1,1039 @@
+	TfwPcntCtl	ctl[TFW_STATS_RANGES];
+	TfwPcntRanges	*pcntrng;
+} TfwApmRBLstEnt;
+
+/*
+ * The ring buffer list structure. 
+ * This is a supporting structure used in the calculation of percentiles.
+ * The list consists of entries of the ring buffer that take part in the
+ * calculation of percentiles. Also it keeps related data that are useful
+ * in making decisions on the recalculation of percentiles.
+ * @rblstent	- The array of ring buffer entries for the calculation.
+ * @rblstsz	- The current size of @rblstent.
+ * @rblstmaxsz	- The maximum size of @rblstent.
+ * @jtmwstamp	- The start of the time window the percentiles are for.
+ * @entry_cnt	- The number of hits in the current buffer ring entry.
+ * @total_cnt	- The number of hits within the current time window.
+ */
+typedef struct {
+	TfwApmRBLstEnt	*rblstent;
+	unsigned int	rblstsz;
+	unsigned int	rblstmaxsz;
+	unsigned long	jtmwstamp;
+	unsigned long	entry_cnt;
+	unsigned long	total_cnt;
+} TfwApmRBLst;
+
+/*
+ * APM Data structure.
+ * Note that the organization of the supporting data heavily depends
+ * on the fact that there's only one party that does the calculation
+ * of percentiles - the function that runs periodically on timer.
+ * If there are several different parties that do the calculation,
+ * then the data may need to be organized differently.
+ * @rbuf	- The ring buffer for the specified time window.
+ * @rblst	- The list of ring buffer entries used in the calculation.
+ * @stats	- The latest percentiles.
+ * @timer	- The periodic timer handle.
+ * @flags	- The atomic flags (see below).
+ */
+#define TFW_APM_DATA_F_REARM	0x0001		/* Re-arm the timer. */
+#define TFW_APM_TIMER_TIMEOUT	(HZ * 3)	/* The timer periodicity. */
+
+typedef struct {
+	TfwApmRBuf		rbuf;
+	TfwApmRBLst		rblst;
+	TfwApmStats		stats;
+	struct timer_list	timer;
+	unsigned long		flags;
+} TfwApmData;
+
+/*
+ * [1ms, 349ms] should be sufficient for almost any installation,
+ * including cross atlantic.
+ */
+static const TfwPcntCtl __read_mostly tfw_rngctl_init[TFW_STATS_RANGES] = {
+	{{0, 1, 16}},
+	{{1, 17, 47}},
+	{{2, 48, 108}},
+	{{4, 109, 349}}
+};
+
+/* A superset of percentiles for all users. */
+const unsigned int __read_mostly tfw_apm_prcntl_ith[] = {
+	1, 50, 75, 90, 95, 99
+};
+
+static int tfw_apm_jtmwindow;		/* Time window in jiffies. */
+static int tfw_apm_jtmintrvl;		/* Time interval in jiffies. */
+static int tfw_apm_tmwscale;		/* Time window scale. */
+
+/*
+ * Get the next bucket in the ring buffer entry that has a non-zero
+ * hits count. Set the bucket's sequential number, the range number,
+ * and the bucket number. Set the response time value for the bucket.
+ */
+/*
+ * Ring buffer entry state structure.
+ * @v	- The response time value.
+ * @i	- The current sequential bucket number across all ranges.
+ * @r	- The current range number.
+ * @b	- The current bucket number.
+ */
+typedef struct {
+	u16	v;
+	u8	i;
+	u8	r : 4;
+	u8	b : 4;
+} __attribute__((packed)) TfwApmRBEState;
```

Comment:
You're renamed `upd` to `update` but still use very 'telling names' like `v`, `i`, `r` and `b` :)